### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by budison**
- Few packages are empty yet but there will happen stuff later
- public void Game.configureGame() will be changed later to not void, so don't attack me! :)
- Could not avoid a few getters... I use them only for testing... Don't know if there is a way to handle that differently without the need of getters

### CODE REVIEW

File: src/main/java/com/github/budison/ArgsReader.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    Player createPlayerByString(String player) {
+        if(player.equalsIgnoreCase("O")) {
+            return new PlayerO("PlayerO", 'O');
+        }
+        else if(player.equalsIgnoreCase("X")) {
+            return new PlayerX("PlayerX", 'X');
+        }
+        else {
+            throw new IllegalArgumentException("startingPlayer or winningPlayer argument is either X nor O!");
```
Comment (deetree) [inline src/main/java/com/github/budison/ArgsReader.java:39]: A lot of imperative code that may be changed into a switch statement

File: src/main/java/com/github/budison/ArgsReader.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            return WinningType.VERTICAL;
+        }
+        else if(winningType.equalsIgnoreCase("diagonal")) {
+            return WinningType.DIAGONAL;
+        }
+        else if(winningType.equalsIgnoreCase("antidiagonal") || winningType.equalsIgnoreCase("anti diagonal")) {
+            return WinningType.ANTI_DIAGONAL;
+        }
+        else {
+            throw new IllegalArgumentException("winningType argument is not valid!");
```
Comment (deetree) [inline src/main/java/com/github/budison/ArgsReader.java]: Switch may be useful here

File: src/main/java/com/github/budison/ArgsReader.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return winningPlayer;
+    }
+
+    WinningType getWinningType() {
+        return winningType;
+    }
+
+    int getWinningLine() {
+        return winningLine;
+    }
```
Comment (deetree) [inline src/main/java/com/github/budison/ArgsReader.java]: GETTERS ARE EVIL, although the fields are not private, so you may access them directly.

File: src/main/java/com/github/budison/LanguageType.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.github.budison;
+
+/**
+ * @author Kevin Nowak
+ */
+public enum LanguageType {
```
Comment (deetree) [inline src/main/java/com/github/budison/LanguageType.java:6]: public

File: src/main/java/com/github/budison/Player.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.github.budison;
+
+/**
+ * @author Kevin Nowak
+ */
+public interface Player {
```
Comment (deetree) [inline src/main/java/com/github/budison/Player.java:6]: public

File: src/main/java/com/github/budison/PlayerO.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package com.github.budison;
+
+/**
+ * @author Kevin Nowak
+ */
+public record PlayerO(String name, char sign) implements Player {}
```
Comment (deetree) [inline src/main/java/com/github/budison/PlayerO.java:6]: public

File: src/main/java/com/github/budison/PlayerX.java [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.github.budison;
+
+/**
+ * @author Kevin Nowak
+ */
+public record PlayerX(String name, char sign) implements Player {}
```
Comment (deetree) [inline src/main/java/com/github/budison/PlayerX.java:6]: public

File: src/main/java/com/github/budison/board/BoardDimension.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
+package com.github.budison.board;
+
+import java.util.Objects;
+
+/**
+ * @author Kevin Nowak
+ */
+public class BoardDimension {
+    int boardDimension;
```
Comment (deetree) [inline src/main/java/com/github/budison/board/BoardDimension.java]: not private

File: src/main/java/com/github/budison/board/GameBoard.java [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package com.github.budison.board;
+
+/**
+ * @author Kevin Nowak
+ */
+public class GameBoard {
```
Comment (deetree) [inline src/main/java/com/github/budison/board/GameBoard.java:6]: public

File: src/main/java/com/github/budison/demo/Demo.java [hunk 10]
```diff
@@ -9,10 +9,10 @@
     public Demo() {
     }
```
Comment (deetree) [inline src/main/java/com/github/budison/demo/Demo.java:10]: default constructor is unnecessary

File: src/main/java/com/github/budison/demo/Demo.java [hunk 11]
```diff
@@ -9,10 +9,10 @@
...
     public Demo(ArgsReader argsReader) {
-
+        System.out.println("Demo Mode with arguments");
     }
 
     public void run() {
```
Comment (deetree) [inline src/main/java/com/github/budison/demo/Demo.java:15]: public void

File: src/main/java/com/github/budison/game/Game.java [hunk 12]
```diff
@@ -1,10 +1,31 @@
...
+    public GameConfigDataHolder gameConfigDataHolder;
+    LanguageType languageType;
+    GameState gameState;
+    GameBoard gameBoard;
+
+    public Game(LanguageType languageType) {
+        this.languageType = languageType;
+    }
+
+    public Game() {}
```
Comment (deetree) [inline src/main/java/com/github/budison/game/Game.java]: empty c-tor

File: src/main/java/com/github/budison/game/Game.java [hunk 13]
```diff
@@ -1,10 +1,31 @@
...
+    public void configureGame() {
+        this.gameConfigDataHolder = new GameConfigurator().configureGame();
+        this.languageType = this.gameConfigDataHolder.languageType();
+    }
+
+    // Getters (only) used for testing purposes
+
+    public LanguageType getLanguageType() {
+        return languageType;
     }
```
Comment (deetree) [inline src/main/java/com/github/budison/game/Game.java]: the same as before, your field is accessible from outside

File: src/main/java/com/github/budison/game/Game.java [hunk 14]
```diff
@@ -1,10 +1,31 @@
...
+    GameState gameState;
+    GameBoard gameBoard;
+
+    public Game(LanguageType languageType) {
+        this.languageType = languageType;
+    }
+
+    public Game() {}
+
+    public void configureGame() {
```
Comment (deetree) [inline src/main/java/com/github/budison/game/Game.java]: void

File: src/main/java/com/github/budison/game/GameConfigurator.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import com.github.budison.PlayerX;
+import com.github.budison.board.BoardDimension;
+import com.github.budison.io.ioController;
+
+/**
+ * @author Kevin Nowak
+ */
+class GameConfigurator {
+    GameConfigDataHolder gameConfigDataHolder;
+    ioController ioController;
```
Comment (deetree) [inline src/main/java/com/github/budison/game/GameConfigurator.java]: pckg-private

File: src/main/java/com/github/budison/io/ioGameConfiguration.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+                }
+            } catch (IllegalArgumentException illegalArgumentException) {
+                System.err.println("Invalid input! Try again.");
+                repeat = true;
+            }
+        }
+        switch (inputInt) {
+            case 2: return LanguageType.DE;
+            default: return LanguageType.EN;
+        }
```
Comment (deetree) [inline src/main/java/com/github/budison/io/ioGameConfiguration.java]: you may use return switch ........

File: src/main/java/com/github/budison/io/ioGameConfiguration.java [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+            String input = in.next();
+            try {
+                inputInt = Integer.parseInt(input);
+                if(inputInt > 2 || inputInt < 1) {
+                    throw new IllegalArgumentException();
+                }
+            } catch (IllegalArgumentException illegalArgumentException) {
+                System.err.println("Invalid input! Try again.");
+                repeat = true;
+            }
```
Comment (deetree) [inline src/main/java/com/github/budison/io/ioGameConfiguration.java]: you can extract the method so this one would be shorter

**Review by deetree — COMMENTED** [review]
It's good that you are creating JavaDocs

File: src/main/java/com/github/budison/game/GameConfigurator.java (hunk 15 above)
Comment (budison) [reply src/main/java/com/github/budison/game/GameConfigurator.java]: Whats the problem with that one?

File: src/main/java/com/github/budison/game/Game.java (hunk 13 above)
Comment (budison) [reply src/main/java/com/github/budison/game/Game.java]: Needed for testing.

**Pull request merged**